### DISCUSSION TITLE
[HELP] Script to Close a NTB Modal

### DISCUSSION THREAD

**Discussion by FelipeRearden** [chrisgurney/obsidian-note-toolbar#355] [Scripting]
Hello @chrisgurney !!!

I am trying to create a templater script where the a NTB Modal will close after a 10000ms. Something like this…

```
<%*
const filename = "Lorem Ipsum Generator.md";
const file = app.vault.getAbstractFileByPath(filename);
await ntb.modal(file,{
title: "LOREM IPSUM"
});
setTimeout(() => {
 \\code to close the modal
}, 10000);
%>
```

I tried to use `.detach` but my code knowledge is not enough to develop this script 🥲

The idea is to transform the NTB Modal in a notification for a specific workflow 🙏 

Do you have idea on how to do this? 
Thank you very much ☀️

**Comment by chrisgurney** [comment]
@FelipeRearden 

> I am trying to create a templater script where the a NTB Modal will close after a 10000ms. Something like this…

I'll update the API to return a reference to `Modal`, since `ntb.modal()` doesn't need to return anything otherwise (it's currently returning `void`). From there you can then call `.close()` on it, like so:

```ts
const modal = await ntb.modal('Goodbye World');
setTimeout(() => {
  modal.close();
}, 2000);
```

(This will be a special case as I'm not returning references to other UI.)

Watch for an update in the next release ([commit](https://github.com/chrisgurney/obsidian-note-toolbar/commit/eb58090224595332961211f37d49a17f443c7d2a)).

**Reply by FelipeRearden** [reply]
Amazing @chrisgurney !!!

Thank you very much for this special treatment. It will open a new workflow for NTB modal for me 🙏 🙏 🙏

I wil test the next release and get to you ☀️
Thanks again for one more feature 🥇

**Reply by chrisgurney** [reply]
@FelipeRearden It's now live in [`1.24.03`](https://github.com/chrisgurney/obsidian-note-toolbar/releases/tag/1.24.03).

**Reply by FelipeRearden** [reply]
Working absolutely perfect on version `1.24.03` !!!!

Thank you very much @chrisgurney ! you are super 🥇

**Comment by FelipeRearden** [comment]
Hello @chrisgurney !!!!

Let me ask you something, would be possible to add the same treatment made for closing a `ntb.modal` to the `ntb.menu` ?

I tried:
```js
const item = await ntb.menu('Daily Notes');
await new Promise(r => setTimeout(r, 1000));
item.close();
```
or

```js
const item = await ntb.menu('Daily Notes');
await new Promise(r => setTimeout(r, 1000));
item.detach();
```

I have a complex scenario for this workflow. if you want I can share with you!

Thanks! I wish you a fantastic day ☀️

**Reply by chrisgurney** [reply]
Hey @FelipeRearden can you please elaborate on your use case?

There's no need to "close" a menu, as it closes automatically when an item is selected.

Maybe a callback function could be provided, to process the selected (custom) menu item? Something like:

```JS
await ntb.menu([
  { type: 'function', value: 'Some value', label: 'Do the thing', icon: 'smile', function: () => { SOME CODE HERE } },
  ...
]);

```

**Reply by FelipeRearden** [reply]
**Chris while I was writing this answer I realize that what I wanna do is not feasible. I made an logic error and the code that I was expecting for** `menu.close()` **is not gonna work the way that I needed**

This WAS the idea 😭

<hr>

Chris, let me explain what I mean by close:

- The video shows a menu called `menu1` that opens when I hit a hotkey.
- **While the menu is open** I hit another key that **closes** `menu1` and opens another menu called `menu2`


https://github.com/user-attachments/assets/f75d76ed-9deb-4a4f-8994-597d0fa18d57

In other words: I am switching between `menu1` and `menu2` using a hotkey

To achieve this I use the plugin [Sequence Hotkeys](https://github.com/moolmanruan/obsidian-sequence-hotkeys) that let  me run a command when a key is hit. My case is <kbd>`</kbd>

This is the command:
```js
if (activeDocument.body.find(".menu.note-toolbar-menu.suggestermenu")) {
activeWindow.activeDocument.querySelector(":focus").dispatchEvent(new KeyboardEvent("keydown", { key: "Escape" }));
await new Promise(r => setTimeout(r, 100));
let menu1 = await ntb.menu('Bookmarks',{
position: "toolbar",
class: "suggestermenu",
focusInMenu: true
});
} else if (activeDocument.body.find(".menu.note-toolbar-menu.suggestermenu2")) {
activeWindow.activeDocument.querySelector(":focus").dispatchEvent(new KeyboardEvent("keydown", { key: "Escape" }));
await new Promise(r => setTimeout(r, 100));
let menu2 = await ntb.menu('Recent Files',{
position: "toolbar",
class: "suggestermenu2",
focusInMenu: true
});
} else {
app.commands.executeCommandById('page-scroll:page-scroll-down')
}
```

As you can see I am using 2 different `class` to target each `ntb.menu` and I am using a keyboard event to close the menu before switching to another menu
```js
activeWindow.activeDocument.querySelector(":focus").dispatchEvent(new KeyboardEvent("keydown", { key: "Escape" }));
```

### What is the problem that I am trying to solve

Look at the video again you will see that while going from `menu1` to `menu2` and `menu2` to `menu1` there is a **glitch** since I am losing and opening two different menus.

**MY IDEA** if we have a `menu.close()` I can remove this glitch by opening opening `menu2` before closing `menu1`. 

<hr> 

Sorry for the misunderstanding ☀️
